prov/shm/src/smr_init.c

```diff
@@ -72,6 +72,48 @@ static void smr_resolve_addr(const char *node, const char *service,
 	(*addr)[*addrlen - 1]  = '\0';
 }
 
+static uint64_t smr_calculate_shm_usage(size_t tx_count, size_t rx_count)
+{
+	size_t total_size;
+	size_t tx_size, rx_size;
+
+	tx_size = roundup_power_of_two(tx_count);
+	rx_size = roundup_power_of_two(rx_count);
+
+	total_size = sizeof(struct smr_region) +
+		     sizeof(struct smr_cmd_queue) + sizeof(struct smr_cmd) * rx_size +
+		     sizeof(struct smr_resp_queue) + sizeof(struct smr_resp) * tx_size +
+		     sizeof(struct smr_inject_pool) + sizeof(struct smr_inject_pool_entry) * rx_size +
+		     sizeof(struct smr_sar_pool) + sizeof(struct smr_sar_pool_entry) * SMR_MAX_PEERS +
+		     sizeof(struct smr_peer_data) * SMR_MAX_PEERS +
+		     NAME_MAX;
+	total_size = roundup_power_of_two(total_size);
+	return total_size * SMR_MAX_PEERS;
```

Comment:
Multiplying by MAX_PEERS doesn't look right.  We don't create the above regions per peer, and the areas that are per peer are already multiplied by MAX_PEERS.

Reply:
This is the total size of shm space we need for all peers (there're at most SMR_MAX_PEERS peers in shm provider). 

Of course, if the apps just create 2 local peers, we are over concerned. But we are calculating the upper limit of shm space here, so that we know if we can handle the extreme scenario if the apps create SMR_MAX_PEERS local peers in shm provider.

Reply:
There needs to be comment in the code about what this is doing, because it's non-intuitive.  The check is assuming a some number of peer processes all want to map pages from the same limited pool.  I would replace MAX_PEERS here with number of CPUs/cores as a more likely number.

Isn't there somewhere else in the code where the shm region is being setup where these same calculations need to be made?

Reply:
> There needs to be comment in the code about what this is doing, because it's non-intuitive. The check is assuming a some number of peer processes all want to map pages from the same limited pool. I would replace MAX_PEERS here with number of CPUs/cores as a more likely number.

Sure, I can add the comment to clarify. 
You meant to update macro SMR_MAX_PEERS with a number of CPUs/cores, or just the one in `return total_size * SMR_MAX_PEERS;`?
> 
> Isn't there somewhere else in the code where the shm region is being setup where these same calculations need to be made?

Yes, it's in smr_create(), and it's per endpoint. https://github.com/ofiwg/libfabric/blob/d21e983832acedbf9adf36679c18091c5516c557/prov/util/src/util_shm.c#L86

Reply:
Please separate out this calculation, so it's not duplicated and gets out of sync.

Reply:
And I'm suggesting it should either be total_size * 2 (minimum number of processes sharing mapping space) or total_size * number_of_cores (realistic maximum number of processes sharing mapping space).

Reply:
Let me summarize the way to proceed which is acceptable to you.

1. separate out this calculation.
2. still use the check smr_shm_space_check() in smr_getinfo()
3. but use total_size * 2 or total_size * numer_of_cores instead (I would use number_of_cores.)

Please let me know if I misunderstood anything.

Reply:
That's correct.